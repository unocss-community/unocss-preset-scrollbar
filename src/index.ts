import type { Preset } from 'unocss'
import { colorResolver, handler } from '@unocss/preset-mini/utils'

const defaultOption: Required<PresetScrollbarDefaultOption> = {
  scrollbarWidth: '8px',
  scrollbarHeight: '8px',
  scrollbarTrackRadius: '4px',
  scrollbarThumbRadius: '4px',
  scrollbarTrackColor: '#f5f5f5',
  scrollbarThumbColor: '#ddd',
  varPrefix: '',
  numberToUnit: value => `${value / 4}rem`,
}

export interface PresetScrollbarDefaultOption {
  /**
   * default scrollbar width
   * @default '8px'
   */
  scrollbarWidth?: string
  /**
   * default scrollbar height
   * @default '8px'
   */
  scrollbarHeight?: string
  /**
   * default scrollbar track radius
   * @default '4px'
   */
  scrollbarTrackRadius?: string
  /**
   * default scrollbar thumb radius
   * @default '4px'
   */
  scrollbarThumbRadius?: string
  /**
   * default scrollbar track background color
   * @default '#f5f5f5'
   */
  scrollbarTrackColor?: string
  /**
   * default scrollbar thumb background color
   * @default '#ddd'
   */
  scrollbarThumbColor?: string
  /**
   * css variable prefix
   * @default ''
   */
  varPrefix?: string
  /**
   * convert number to unit
   * @default value => `${value / 4}rem`
   * @example
   * numberToUnit: value => `${value / 4}rem`
   * p-4 => padding: 1rem
   * p-4px => padding: 4px
   *
   * @example
   * numberToUnit: value => `${value}rpx`
   * p-4 => padding: 4rpx
   */
  numberToUnit?: (value: number) => string
}

const customRules = {
  'radius': ['track-radius', 'thumb-radius'],
  'w': ['width'],
  'h': ['height'],
  'track-radius': ['track-radius'],
  'thumb-radius': ['thumb-radius'],
}

const numberVarRegex = new RegExp(`^scrollbar-(${Object.keys(customRules).join('|')})-(\\d+?)([a-zA-Z]*?)$`)

export function presetScrollbar(option?: PresetScrollbarDefaultOption): Preset {
  const config = {
    ...defaultOption,
    ...option,
  }

  function resolveVar(name: string) {
    const prefix = config.varPrefix
    return `--${prefix ? `${prefix}-` : ''}scrollbar-${name}`
  }

  const variantsRE = /^(scrollbar(-track|-thumb)?):.+$/

  return {
    name: 'unocss-preset-scrollbar',
    shortcuts: [
      [
        'scrollbar', [
          { overflow: 'auto' },
          'scrollbar-custom-property',
          `scrollbar-color-[var(${resolveVar('thumb')})_var(${resolveVar('track')})]`,
          `scrollbar-track:scrollbar-background-color-[var(${resolveVar('track')})]`,
          `scrollbar-thumb:scrollbar-background-color-[var(${resolveVar('thumb')})]`,
          `scrollbar:scrollbar-width-[var(${resolveVar('width')})]`,
          `scrollbar:scrollbar-height-[var(${resolveVar('height')})]`,
        ],
      ],
      [
        'scrollbar-rounded', `
          scrollbar-track:scrollbar-border-radius-[var(${resolveVar('track-radius')})]
          scrollbar-thumb:scrollbar-border-radius-[var(${resolveVar('thumb-radius')})]
        `,
      ],
    ],
    variants: [
      // ::-webkit-scrollbar-track
      // ::-webkit-scrollbar-thumb
      // ::-webkit-scrollbar
      (matcher) => {
        if (!variantsRE.test(matcher))
          return

        const variant = matcher.replace(variantsRE, '$1')

        return {
          matcher: matcher.slice(variant.length + 1),
          selector: (s) => {
            return `${s}::-webkit-${variant}`
          },
        }
      },
    ],
    rules: [
      [
        /^scrollbar-color-(.+)$/,
        ([_, prop]) => ({
          'scrollbar-color': handler.bracket.cssvar.auto.fraction.rem(prop),
        }),
      ],
      [
        /^scrollbar-custom-property$/,
        ([_]) => ({
          [resolveVar('track')]: config.scrollbarTrackColor,
          [resolveVar('thumb')]: config.scrollbarThumbColor,
          [resolveVar('width')]: config.scrollbarWidth,
          [resolveVar('height')]: config.scrollbarHeight,
          [resolveVar('track-radius')]: config.scrollbarTrackRadius,
          [resolveVar('thumb-radius')]: config.scrollbarThumbRadius,
        }),
      ],
      [
        /^scrollbar-thumb-color-(.+)$/,
        colorResolver(resolveVar('thumb'), 'scrollbar-thumb'),
        { autocomplete: 'scrollbar-thumb-color-$colors' },
      ],
      [
        /^scrollbar-track-color-(.+)$/,
        colorResolver(resolveVar('track'), 'scrollbar-track'),
        { autocomplete: 'scrollbar-track-color-$colors' },
      ],
      [
        /^scrollbar-(width|height|background\-color|border\-radius|)-(\[var.+\])$/,
        ([_, prop, value]) => {
          return {
            [`${prop}`]: handler.bracket(value),
          }
        },
      ],
      [
        numberVarRegex,
        ([_, type, value, unit]) => {
          const val = unit ? value + unit : config.numberToUnit(Number.parseInt(value))
          const vars = customRules[type as keyof typeof customRules]
            .map(v => resolveVar(v))
          return vars.reduce((acc: any, v) => {
            acc[v] = val
            return acc
          }, {})
        },
        { autocomplete: `scrollbar-(${Object.keys(customRules).join('|')})-<num>` },
      ],
    ],
  }
}
