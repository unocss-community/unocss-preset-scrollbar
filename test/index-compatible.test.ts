import { describe, expect, it } from 'vitest'
import { createGenerator } from 'unocss'
import { presetScrollbar } from '../src'

describe('scrollbar (compatible)', async () => {
  const generator = await createGenerator({
    presets: [
      presetScrollbar({
        noCompatible: false,
      }),
    ],
  })

  it('scrollbar-auto', async () => {
    const { css } = await generator.generate('scrollbar')
    expect(css).toMatchSnapshot()
  })

  it('scrollbar-thin', async () => {
    const { css } = await generator.generate([
      'scrollbar',
      'scrollbar-thin',
    ])
    expect(css).toMatchSnapshot()
  })

  it('scrollbar-none', async () => {
    const { css } = await generator.generate([
      'scrollbar',
      'scrollbar-none',
    ])
    expect(css).toMatchSnapshot()
  })

})
