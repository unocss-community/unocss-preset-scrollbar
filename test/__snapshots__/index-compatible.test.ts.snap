// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scrollbar (compatible) > scrollbar-auto 1`] = `
"/* layer: shortcuts */
.scrollbar{scrollbar-color:var(--scrollbar-thumb) var(--scrollbar-track);scrollbar-width:auto;--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}"
`;

exports[`scrollbar (compatible) > scrollbar-none 1`] = `
"/* layer: shortcuts */
.scrollbar{scrollbar-color:var(--scrollbar-thumb) var(--scrollbar-track);scrollbar-width:auto;--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar-none{scrollbar-width:none;}
.scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}
.scrollbar-none::-webkit-scrollbar{display:none;}"
`;

exports[`scrollbar (compatible) > scrollbar-thin 1`] = `
"/* layer: shortcuts */
.scrollbar{scrollbar-color:var(--scrollbar-thumb) var(--scrollbar-track);scrollbar-width:auto;--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar-thin{scrollbar-width:thin;--scrollbar-width:8px;--scrollbar-height:8px;}
.scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}"
`;
